### PR TITLE
[WIP] New macOS sandbox based on EndpointSecurity APIs

### PR DESCRIPTION
**Pull request by Kr1sso**
Initial draft for EndpointSecurity based sandbox support on macOS Catalina and higher. This is here to test and prototype a new sandbox for the upcoming OS release as kernel extensions are deprecated and will soon be forbidden.

**Important**

This currently only runs on systems with SIP disabled, using super user privileges and signing the bxl binary with a valid developer certificate before executing!

```codesign --entitlements Private/macOS/sandbox.entitlements Out/Bin/Debug/osx-x64/bxl -s XYZ```

A valid signing key (replace XYZ) can be obtained by using a registered MS developer account connected with an AppleID.

Running ```security find-identity -v -p codesigning``` will list those if any are available on the system.

### CODE REVIEW

File: Public/Src/Engine/Processes/SandboxConnectionES.cs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            if (m_esConnectionInfo.Error != Sandbox.SandboxSuccess)
+            {
+                throw new BuildXLException($@"Unable to connect to EndpointSecurity sandbox (Code: {m_esConnectionInfo.Error})");
+            }
+
+#if DEBUG
+            ProcessUtilities.SetNativeConfiguration(true);
+#else
+            ProcessUtilities.SetNativeConfiguration(false);
+
```
Comment (almiliMSFT) [inline Public/Src/Engine/Processes/SandboxConnectionES.cs]:
```suggestion
```

File: Public/Src/Engine/Processes/SandboxConnectionES.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            MeasureCpuTimes = false;
+            IsInTestMode = false;
+
+            var process = System.Diagnostics.Process.GetCurrentProcess();
+
+            Sandbox.InitializeEndpointSecuritySandbox(ref m_esConnectionInfo, process.Id);
+            if (m_esConnectionInfo.Error != Sandbox.SandboxSuccess)
+            {
+                throw new BuildXLException($@"Unable to connect to EndpointSecurity sandbox (Code: {m_esConnectionInfo.Error})");
```
Comment (almiliMSFT) [inline Public/Src/Engine/Processes/SandboxConnectionES.cs]:
```suggestion
                throw new BuildXLException($"Unable to connect to EndpointSecurity sandbox (Code: {m_esConnectionInfo.Error})");
```

File: Public/Src/Engine/Processes/SandboxConnectionES.cs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                }
+            };
+
+            Sandbox.ObserverFileAccessReports(ref m_esConnectionInfo, callback, Marshal.SizeOf<Sandbox.AccessReport>());
+        }
+
+        /// <inheritdoc />
+        public bool NotifyUsage(uint cpuUsage, uint availableRamMB)
+        {
+            // TODO: Will we need this?
```
Comment (almiliMSFT) [inline Public/Src/Engine/Processes/SandboxConnectionES.cs]: Nope.  The kext uses this stuff for dynamic throttling, which in the end was not very useful

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.cpp [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+
+    SandboxedProcess *newRecord = record;
+    if (newRecord == nullptr)
+    {
+        return kTrieResultAlreadyExists;
+    }
+    
+    if (node->record_ != nullptr)
+    {
+//        delete newRecord;
```
Comment (Kr1sso) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.cpp]: For debugging this is currently disabled - will address later.

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.cpp [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+    {
+        return kTrieResultFailure;
+    }
+
+    SandboxedProcess *previousValue = node->record_;
+    
+    if (previousValue != nullptr)
+    {
+        node->record_ = (SandboxedProcess *) value;
+//        delete previousValue;
```
Comment (Kr1sso) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.cpp]: For debugging this is currently disabled - will address later.

File: Public/Src/Utilities/Interop/MacOS/Sandbox.cs [hunk 6]
```diff
@@ -79,15 +112,26 @@ public static void InitializeKextSharedMemory(KextConnectionInfo connectionInfo,
...
+            
+            /// <nodoc />
+            EndpointSecurity
+        }
+
         /// <nodoc />
         [DllImport(Libraries.BuildXLInteropLibMacOS, EntryPoint = "SendPipStarted")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        public static extern bool SendPipStarted(int processId, long pipId, byte[] famBytes, int famBytesLength, KextConnectionInfo info);
+        public static extern bool SendPipStarted(int processId, long pipId, byte[] famBytes, int famBytesLength, ConnectionType type, ref KextConnectionInfo info);
```
Comment (almiliMSFT) [inline Public/Src/Utilities/Interop/MacOS/Sandbox.cs]: Why `ref` for `KextConnectionInfo`?

File: Public/Src/Engine/Scheduler/Scheduler.cs [hunk 7]
```diff
@@ -4760,12 +4764,12 @@ protected virtual bool InitSandboxedKextConnection(LoggingContext loggingContext
                                 uint availableRamMB = double.IsNaN(availableRam) || double.IsInfinity(availableRam)
                                     ? 0
                                     : Convert.ToUInt32(Math.Round(availableRam));
-                                kextConnection.NotifyUsage(cpuUsageBasisPoints, availableRamMB);
+                                sandboxConnection.NotifyUsage(cpuUsageBasisPoints, availableRamMB);
                             };
-                        }
+                        }
```
Comment (almiliMSFT) [inline Public/Src/Engine/Scheduler/Scheduler.cs]:
```suggestion
                        } 
```

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/SandboxedPip.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+ * Represents the root of the process tree being tracked.
+ *
+ * The 'Pip' name comes from the BuildXL terminology, where 'pip' is a generic build task
+ * that may spawn arbitrary number of child processes.
+ *
+ * Every pip comes with a 'FileAccessManifest' (FAM).  A FAM contains all the policies relevant
+ * for sandboxing a pip, e.g., which file accesses are permitted, which are not, which should
+ * be reported back, etc.
+ */
+class SandboxedPip
```
Comment (almiliMSFT) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/SandboxedPip.hpp]: Is it safe to assume that nothing really changed in this class?

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/SandboxedProcess.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    if (pip_ == nullptr)
+    {
+        throw "No valid SandboxedPip provided on SandboxedProcess construction!";
+    }
+}
+
+SandboxedProcess::~SandboxedProcess()
+{
+    print_backtrace();
```
Comment (almiliMSFT) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/SandboxedProcess.cpp]:
```suggestion
    // print_backtrace();
```

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/SandboxedProcess.hpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    inline pid_t getPid() const                                 { return id_; }
+
+    /*! Returns whether a full path has been set */
+    inline bool hasPath() const                                 { return path_[0] == '/'; }
+
+    /*! 0-terminated full path to the executable file of this process */
+    inline const char* getPath() const                          { return path_; }
+
+    /*! Copies the 0-terminated string in 'path' to its own path buffer. */
+    inline void setPath(const char *path, size_t len = PATH_MAX)   { strlcpy(path_, path, len); }
```
Comment (almiliMSFT) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/SandboxedProcess.hpp]:
```suggestion
    inline void setPath(const char *path, size_t len = PATH_MAX){ strlcpy(path_, path, len); }
```

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.hpp [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+};
+
+// ================================== class Trie ==================================
+
+/*!
+ * A thread-safe, lock-free, dictionary implementation.
+ *
+ * Only 2 types of keys are allowed: (1) an unsigned integer, and (2) an ascii path.
+ *
+ * A value must be a pointer to an arbitrary OSObject.  Once an OSObject is added
```
Comment (almiliMSFT) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/Data/Trie.hpp]: These comments about `OSObject` and `OSDictionary` are probably not relevant any more (just noticing, nothing needs to be done about it now)

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/ESSandbox.cpp [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+
+#include <signal.h>
+#include "ESSandbox.h"
+#include "BuildXLSandboxShared.hpp"
+
+#include "IOHandler.hpp"
+
+static ESSandbox *sandbox;
+
+dispatch_queue_t serialQueue = dispatch_queue_create("com.blah.queue", DISPATCH_QUEUE_SERIAL);
```
Comment (almiliMSFT) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/ESSandbox.cpp]:
```suggestion
dispatch_queue_t serialQueue = dispatch_queue_create("com.microsoft.buildxl.queue", DISPATCH_QUEUE_SERIAL);
```

File: Public/Src/Sandbox/MacOs/Interop/Sandbox/ESSandbox.cpp [hunk 13]
```diff
@@ -0,0 +1,354 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <signal.h>
+#include "ESSandbox.h"
+#include "BuildXLSandboxShared.hpp"
+
+#include "IOHandler.hpp"
+
+static ESSandbox *sandbox;
```
Comment (almiliMSFT) [inline Public/Src/Sandbox/MacOs/Interop/Sandbox/ESSandbox.cpp]: maybe add some comment about where this gets initialized

File: Public/Src/Engine/Processes/SandboxConnectionES.cs [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+        /// <inheritdoc />
+        public ulong MinReportQueueEnqueueTime => Volatile.Read(ref m_reportQueueLastEnqueueTime);
+
+        /// <inheritdoc />
+        public bool IsInTestMode { get; }
+
+        private readonly ConcurrentDictionary<long, SandboxedProcessMac> m_pipProcesses = new ConcurrentDictionary<long, SandboxedProcessMac>();
+
+        private Sandbox.KextConnectionInfo m_fakeKextConnectionInfo = new Sandbox.KextConnectionInfo();
```
Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxConnectionES.cs]: >m_fakeKextConnectionInfo [](start = 43, length = 24)

What gives? Is this a todo to refactor away in the future?

File: Public/Src/Engine/Processes/SandboxConnectionES.cs [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+
+#if DEBUG
+            ProcessUtilities.SetNativeConfiguration(true);
+#else
+            ProcessUtilities.SetNativeConfiguration(false);
+            
+#endif
+
+            m_workerThread = new Thread(() => StartReceivingAccessReports());
+            m_workerThread.IsBackground = true;
```
Comment (mpysson) [inline Public/Src/Engine/Processes/SandboxConnectionES.cs]: >m_workerThread [](start = 12, length = 14)

Suggest adding a thread

m_workerTread.Name = "EndpointSecurityDrainer"

**Comment by wjk** [comment]
Excellent work! A few thoughts I had:

* This implementation doesn't actually block access to anything, only log things. Is this planned?
* Would it be possible to move the ES client to a secondary process, so we won't have to run `bxl` as root every time? It could be a setuid tool installed in a secure location that exits when `bxl` is done, or an always-running background daemon that `bxl` communicates with via some IPC mechanism.
* I don't know if Apple will approve the use of the ES entitlement for production deployment (which would remove the requirement to disable SIP), but it's worth an ask. Would this be possible? If so, we could do what [Santa](https://github.com/google/santa) does and distribute the part which must be specially signed pre-compiled, and allow local modifications of the rest. Upon merged PRs for the signed component, Microsoft would then create another signed release at their convenience.

Thanks!

**Comment by almiliMSFT** [comment]
#704

**Comment by Kr1sso** [comment]
> Excellent work! A few thoughts I had:
> 
> * This implementation doesn't actually block access to anything, only log things. Is this planned?
> * Would it be possible to move the ES client to a secondary process, so we won't have to run `bxl` as root every time? It could be a setuid tool installed in a secure location that exits when `bxl` is done, or an always-running background daemon that `bxl` communicates with via some IPC mechanism.
> * I don't know if Apple will approve the use of the ES entitlement for production deployment (which would remove the requirement to disable SIP), but it's worth an ask. Would this be possible? If so, we could do what [Santa](https://github.com/google/santa) does and distribute the part which must be specially signed pre-compiled, and allow local modifications of the rest. Upon merged PRs for the signed component, Microsoft would then create another signed release at their convenience.
> 
> Thanks!

Thanks for your feedback, its noted and appreciated! In regards to your suggestions:

* Our sandbox never has been really declining file accesses and BuildXL itself fails build pips if file access violations are recorded. There are some TODO comments in the ES sandbox code that mention that auth could be potentially implemented in some places. Currently we don't plan to do so.

* Apples plan for EndpointSecurity is to provide it as a bundled extension via SystemExtensions (see https://developer.apple.com/system-extensions/). This will work in a way that does not mandate SIP to be disabled, will prompt the user on first execution and ask for permission. Hence no additional process is required and this implementation is just what is currently possible with the latest Catalina beta. As soon as Apple release better tooling support, we'll adjust our code accordingly.

* Our macOS binary with the ES sandbox will be signed and notarized for distribution once all things are ready for prime time.

**Pull request closed without merging**